src/Core/tests/UnitTests/ImageSource/ImageSourceToImageSourceServiceTypeMappingTests.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Maui.Hosting;
+using Xunit;
+
+namespace Microsoft.Maui.UnitTests.ImageSource
+{
+	[Category(TestCategory.Core, TestCategory.ImageSource)]
+	public class ImageSourceToImageSourceServiceTypeMappingTests
+	{
+		[Fact]
+		public void FindsCorrespondingImageSourceServiceType()
+		{
+			var mapping = new ImageSourceToImageSourceServiceTypeMapping();
+			mapping.Add<IStreamImageSource, IImageSourceService<IStreamImageSource>>();
+			mapping.Add<IUriImageSource, IImageSourceService<IUriImageSource>>();
+
+			var type = mapping.FindImageSourceServiceType(typeof(StreamImageSourceStub));
+
+			Assert.Equal(typeof(IImageSourceService<IStreamImageSource>), type);
+		}
+
+		[Fact]
+		public void ThrowsWhenThereIsNoMatchingMapping()
+		{
+			var mapping = new ImageSourceToImageSourceServiceTypeMapping();
+			mapping.Add<StreamImageSourceStub, IImageSourceService<StreamImageSourceStub>>();
+
+			Assert.Throws<InvalidOperationException>(() => mapping.FindImageSourceServiceType(typeof(IStreamImageSource)));
+		}
+
+		[Theory]
+		[InlineData(typeof(IFileImageSource), typeof(IImageSourceService<IFileImageSource>))]
+		[InlineData(typeof(MyCustomImageSource), typeof(IImageSourceService<IFileImageSource>))]
+		[InlineData(typeof(MyLargeCustomImageSource), typeof(IImageSourceService<MyLargeCustomImageSource>))]
+		[InlineData(typeof(MySmallCustomImageSource), typeof(IImageSourceService<MySmallCustomImageSource>))]
+		public void FindsClosestApplicableMapping(Type type, Type expectedImageSourceServiceType)
+		{
+			var mapping = new ImageSourceToImageSourceServiceTypeMapping();
+			mapping.Add<IFileImageSource, IImageSourceService<IFileImageSource>>();
+			mapping.Add<MyLargeCustomImageSource, IImageSourceService<MyLargeCustomImageSource>>();
+			mapping.Add<MySmallCustomImageSource, IImageSourceService<MySmallCustomImageSource>>();
+
+			var imageSourceServiceType = mapping.FindImageSourceServiceType(type);
+
+			Assert.Equal(expectedImageSourceServiceType, imageSourceServiceType);
+		}
+
+		[Fact]
+		public void FindsMostDerivedBaseInterface()
+		{
+			var mapping = new ImageSourceToImageSourceServiceTypeMapping();
+			mapping.Add<IFileImageSource, IImageSourceService<IFileImageSource>>();
+			mapping.Add<IMyCustomImageSource, IImageSourceService<IMyCustomImageSource>>();
+
+			var imageSourceServiceType = mapping.FindImageSourceServiceType(typeof(MyCustomImageSource));
+
+			Assert.Equal(typeof(IImageSourceService<IMyCustomImageSource>), imageSourceServiceType);
+		}
+
+		[Fact]
+		public void InCaseOfAmbiguousMatchSelectsTheFirstRegisteredService()
```

Comment:
Is this the right behavior? I think in the original system, it would match the first interface in the list `.GetInterfaces()`, which in this case would be `IFileImageSource`?

Reply:
You're right. This would behave differently from the previous implementation. I wonder if we should throw `InvalidOperationException` in this case because there isn't a unique match 🤔 I'll sleep on it 😄